cli/azd/cmd/env.go

```diff
@@ -150,6 +168,293 @@ func (e *envSetAction) Run(ctx context.Context) (*actions.ActionResult, error) {
+	}
+	// prompt for vault selection
+	e.console.StopSpinner(ctx, "", input.Step)
+
+	atLeastOneKvAccountExists := len(vaultsList) > 0
+	if !atLeastOneKvAccountExists && !willCreateNewSecret {
+		e.console.MessageUxItem(ctx, &ux.WarningMessage{
+			Description: "No Key Vaults accounts were found in the selected subscription",
+		})
+		// update the flow to offer creating a new Key Vault
+		willCreateNewSecret = true
+	}
+
+	createNewKvAccountOption := "Create a new Key Vault account"
+	selectKvAccountOptions := []string{}
+	// indexOffset makes the ids to start from 1 instead of 0 when displaying the options
+	indexOffset := 1
+	if willCreateNewSecret {
+		selectKvAccountOptions = append(selectKvAccountOptions, createNewKvAccountOption)
+	}
+	for index, vault := range vaultsList {
+		selectKvAccountOptions = append(selectKvAccountOptions, fmt.Sprintf("%2d. %s", index+indexOffset, vault.Name))
+	}
+
+	kvAccountSelectionIndex, err := e.console.Select(ctx, input.ConsoleOptions{
+		Message:      pickKvAccount,
+		Options:      selectKvAccountOptions,
+		DefaultValue: selectKvAccountOptions[0],
+	})
+	if err != nil {
+		return nil, fmt.Errorf("selecting Key Vault account: %w", err)
+	}
+
+	willCreateNewKvAccount := selectKvAccountOptions[kvAccountSelectionIndex] == createNewKvAccountOption
+	if willCreateNewSecret && !willCreateNewKvAccount {
+		// when willCreateNewSecret is true, we added a new option at the beginning of the list
+		// to recover the original kv account name
+		kvAccountSelectionIndex--
+	}
+
+	var kvAccount keyvault.Vault
+	if atLeastOneKvAccountExists {
+		kvAccount = vaultsList[kvAccountSelectionIndex]
+	}
+
+	if willCreateNewKvAccount {
+		location, err := e.prompter.PromptLocation(ctx, subId, "Select the location to create the Key Vault account", nil)
+		if err != nil {
+			return nil, fmt.Errorf("prompting for Key Vault location: %w", err)
+		}
+		rg, err := e.prompter.PromptResourceGroupFrom(ctx, subId, location, prompt.PromptResourceGroupFromOptions{
+			DefaultName: "rg-for-my-kv-account",
+		})
+		if err != nil {
+			return nil, fmt.Errorf("prompting for resource group: %w", err)
+		}
+
+		for {
+			kvAccountName, err := e.console.Prompt(ctx, input.ConsoleOptions{
+				Message: "Enter a name for the Key Vault account",
+				Help:    "The name must be unique within the subscription and must be between 3 and 24 characters long",
+			})
+			if err != nil {
+				return nil, fmt.Errorf("prompting for Key Vault account name: %w", err)
+			}
+			if kvAccountName == "" {
+				e.console.Message(ctx, "Key Vault account name cannot be empty")
+				continue
+			}
+			e.console.ShowSpinner(ctx, "Creating Key Vault account", input.Step)
+			vault, err := e.kvService.CreateVault(ctx, tenantId, subId, rg, location, kvAccountName)
+			e.console.StopSpinner(ctx, "", input.Step)
+			if err != nil {
+				e.console.Message(ctx, fmt.Sprintf("Error creating Key Vault account: %v", err))
+				continue
```

Comment:
Does it make sense to separate out errors before re-prompting? I can imagine if the user doesn't have a permission to create KeyVault for example, this would end up re-prompting them.